src/java/org/apache/cassandra/db/virtual/PrimaryIdTable.java

```diff
@@ -0,0 +1,350 @@
+    {
+        final KeyReader reader;
+        try
+        {
+            // ignore warning on try-with-resources, the reader will be closed on endOfData or close
+            reader = sstable.keyReader(range.left);
+        }
+        catch (IOException e)
+        {
+            logger.error("Error generating keyReader for SSTable: {}", sstable, e);
+            throw new RuntimeException(e);
+        }
+
+        return new AbstractUnfilteredRowIterator(metadata, partitionKey, DeletionTime.LIVE,
+                                                 metadata.regularAndStaticColumns(), Rows.EMPTY_STATIC_ROW,
+                                                 false, EncodingStats.NO_STATS)
+        {
+            public Unfiltered endOfData()
+            {
+                reader.close();
+                return super.endOfData();
+            }
+
+            public void close()
+            {
+                reader.close();
+            }
+
+            private Row buildRow(Clustering<?> clustering, long size)
+            {
+                Row.Builder row = BTreeRow.sortedBuilder();
+                row.newRow(clustering);
+                row.addCell(cell(sizeEstimateColumn, CounterContext.instance().createUpdate(size)));
+                row.addCell(cell(sstablesColumn, CounterContext.instance().createUpdate(1)));
+                return row.build();
+            }
+
+            @Override
+            protected Unfiltered computeNext()
+            {
+                while (!reader.isExhausted())
+                {
+                    DecoratedKey key = target.partitioner.decorateKey(reader.key());
+
+                    for (Consumer<DecoratedKey> listener : readListener)
+                        listener.accept(key);
+
+                    // Store the reader's current data position to calculate size later
+                    long lastPosition = reader.dataPosition();
+                    try
+                    {
+                        // Advance the reader to the next key for the next iteration. Also by moving to next key
+                        // we move the dataPosition to the start of the next key for calculating size
+                        reader.advance();
+                    }
+                    catch (IOException e)
+                    {
+                        logger.error("Error advancing reader for SSTable: {}", sstable, e);
+                        return endOfData();
+                    }
+
+                    // Calculate the size of the current key. If EOF use the length of the file
+                    long current = reader.dataPosition() == -1 ? sstable.uncompressedLength() : reader.dataPosition();
+                    long size = current - lastPosition;
+
+                    String keyString = target.partitionKeyType.getString(key.getKey());
```

Comment:
```suggestion
                    String keyString = target.partitionKeyType.asCQL3Type().toCQLLiteral(key.getKey());
```

using cql literals is far better than type strings

Reply:
This is actually quite done deliberately. If you apply that suggestion you will see the tests start to fail.

Ill add a comment explaining it more  and try to do it better than rambling here, but the problem is if you use CQL Literal for a blob for example you get "0x1232..." but then if you query it for key = "0x123..." it will fail to parse because the AbstractTypes 'fromString' logic only handles the types, not the CQL Literal types. So for `target.partitionKeyType.fromString(keyString);` to work with the same input as is outputted in the tables we need the rawTypes not the literals.

Reply:
actually... I guess I can .asCQLType().fromCQLiteral() that i missed when ran into this originally. Ill update it!